[master] refactor: issue #65604

abismor opened the pull request:
### What does this PR do?
replace deprecated datetime.utcnow() with datetime.now(tz=timezone.utc)

### What issues does this PR fix or reference?
Fixes: #[65604](https://github.com/saltstack/salt/issues/65604) 

### Commits signed with GPG?
No

### Self comment: 
This is my first issue working on someone else's repo.  Althou trivial I hope I've followed correct procedure.

lkubb commented:
There's many more instances that should be replaced:

```console
$ rg -l datetime.utcnow
setup.py
salt/state.py
salt/client/ssh/__init__.py
salt/modules/win_timezone.py
salt/modules/vsphere.py
salt/beacons/status.py
salt/modules/x509_v2.py
salt/modules/purefb.py
salt/modules/tls.py
salt/modules/x509.py
salt/modules/inspectlib/fsdb.py
salt/modules/system.py
salt/modules/purefa.py
salt/grains/core.py
salt/states/boto_iot.py
salt/states/boto_dynamodb.py
salt/states/x509_v2.py
salt/states/github.py
salt/cloud/clouds/ec2.py
salt/cloud/clouds/joyent.py
salt/utils/aws.py
salt/utils/versions.py
salt/utils/jid.py
salt/utils/x509.py
salt/utils/timeutil.py
salt/utils/event.py
salt/spm/pkgdb/sqlite3.py
tests/pytests/unit/utils/test_aws.py
tests/pytests/functional/modules/test_system.py
tests/unit/modules/test_x509.py
tests/unit/test_fileserver.py
tools/pkg/repo/create.py
tools/changelog.py
tools/utils/repo.py
tools/vm.py
```

azuro0092 commented:
My apologies, I've realised flaw of my approach - hence the revert. I will update soon.

azuro0092 commented:
I'm confident this is now done correctly. Thank You for Your time to review my commit.

dwoz commented:
@azuro0092 Looks like there are conflicts again. Pre-commit and lint also failed.

leeclemens commented:
It seems a lot of the Pre-Commit failures are not related to the changes:
``` diff
-                data = __salt__["status.{}".format(func)]()
+                data = __salt__[f"status.{func}"]()
```

Oddly enough, the style-guides that I've found (SaltStack and linked Google guides) show `.format()` is acceptable.

lkubb commented on salt/modules/x509_v2.py:
```diff
@@ -1374,7 +1374,7 @@ def expires(certificate, days=0):
     """
     cert = x509util.load_cert(certificate)
     # dates are encoded in UTC/GMT, they are returned as a naive datetime object
-    return cert.not_valid_after <= datetime.datetime.utcnow() + datetime.timedelta(
```
This will fail because `cryptography` returns naive datetimes. They should be converted to UTC. Once there is a release with https://github.com/pyca/cryptography/issues/9186, this should be updated to try to use the new endpoints, otherwise fallback to the old one with the mentioned fix.

lkubb commented on salt/modules/x509_v2.py:
```diff
@@ -1374,7 +1374,7 @@ def expires(certificate, days=0):
     cert = x509util.load_cert(certificate)
     # dates are encoded in UTC/GMT, they are returned as a naive datetime object
-    return cert.not_valid_after <= datetime.datetime.utcnow() + datetime.timedelta(
+    return cert.not_valid_after <= datetime.now(tz=timezone.utc) + timedelta(
```
This will fail because `cryptography` returns naive datetimes. They should be converted to UTC. Once there is a release with https://github.com/pyca/cryptography/issues/9186, this should be updated to try to use the new endpoints, otherwise fallback to the old one with the mentioned fix.

twangboy reviewed (changes requested):
This needs a changelog... and a rebase.

abismor committed:
Replace deprecated datetime.utcnow() with timezone-aware datetime.now(tz=...)

Resolves issue #65604

Python 3.12 deprecated datetime.utcnow() (and datetime.utcfromtimestamp()).
This replaces all internal call sites with the timezone-aware equivalent
datetime.now(tz=datetime.timezone.utc) so the DeprecationWarning is cleared.

For files that imported the module with 'import datetime', the calls are
rewritten as 'datetime.datetime.now(tz=datetime.timezone.utc)' without
changing the import — this keeps every other 'datetime.X' reference in
those files valid (e.g. datetime.timedelta, datetime.datetime.strptime).

For files that imported with 'from datetime import datetime, ...', the
import line gains ', timezone' and the calls become
'datetime.now(tz=timezone.utc)'.

salt/utils/timeutil.py intentionally wraps the deprecated method and is
left untouched.
